### ISSUE TITLE
Test `filter_with_strategy_and_map` fails seemingly on 32-bit platforms

### ISSUE THREAD

**Issue by nc7s** [frozenlib/test-strategy#13]
Hi, I packaged this in Debian, and noticed that it [failed](https://tracker.debian.org/pkg/rust-test-strategy) the same titular test on the [armel](https://ci.debian.net/packages/r/rust-test-strategy/testing/armel/47732237/), [armhf](https://ci.debian.net/packages/r/rust-test-strategy/testing/armhf/47732162/), and [i386](https://ci.debian.net/packages/r/rust-test-strategy/testing/i386/47444944/) platforms (check links for detailed test log). This pattern occurred to me as 32-bit only, because these are 32-bit, while all passing platforms are 64-bit.

```
(...)
150s test strategy_tuple_struct ... ok
150s test strategy_struct ... ok
150s test strategy_rank3 ... ok
150s test filter_with_strategy_and_map ... FAILED
150s 
150s failures:
150s 
150s ---- filter_with_strategy_and_map stdout ----
150s thread 'filter_with_strategy_and_map' panicked at tests/test_helpers.rs:36:48:
150s called `Result::unwrap()` on an `Err` value: Reason("Too many local rejects")
150s stack backtrace:
150s    0: rust_begin_unwind
150s              at /usr/src/rustc-1.75.0/library/std/src/panicking.rs:645:5
150s    1: core::panicking::panic_fmt
150s              at /usr/src/rustc-1.75.0/library/core/src/panicking.rs:72:14
150s    2: core::result::unwrap_failed
150s              at /usr/src/rustc-1.75.0/library/core/src/result.rs:1653:5
150s    3: core::result::Result<T,E>::unwrap
150s              at /usr/src/rustc-1.75.0/library/core/src/result.rs:1077:23
150s    4: arbitrary::test_helpers::assert_eq_strategy_ops
150s              at ./tests/test_helpers.rs:36:22
150s    5: arbitrary::test_helpers::assert_eq_strategy
150s              at ./tests/test_helpers.rs:25:5
150s    6: arbitrary::test_helpers::assert_arbitrary
150s              at ./tests/test_helpers.rs:13:5
150s    7: arbitrary::filter_with_strategy_and_map
150s              at ./tests/arbitrary.rs:1627:5
150s    8: arbitrary::filter_with_strategy_and_map::{{closure}}
150s              at ./tests/arbitrary.rs:1596:34
150s    9: core::ops::function::FnOnce::call_once
150s              at /usr/src/rustc-1.75.0/library/core/src/ops/function.rs:250:5
150s   10: core::ops::function::FnOnce::call_once
150s              at /usr/src/rustc-1.75.0/library/core/src/ops/function.rs:250:5
150s note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
150s 
150s 
150s failures:
150s     filter_with_strategy_and_map
150s 
150s test result: FAILED. 102 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.35s
150s 
150s error: test failed, to rerun pass `--test arbitrary`
150s autopkgtest [09:30:40]: test rust-test-strategy:@: -----------------------]
```

With a few `eprintln!`s, I managed to get these:

- `reject_local` is triggered in `proptest::strategy::filter::Filter::new_tree()`
- `source: Just(29)`
- `whence: Reason("_y % 2 == 0")`

Unfortunately I'm unfamiliar with arbitrary and proptest, thus can't dig much deeper.

**Comment by frozenlib** [comment]
Thank you for your report.

I was able to confirm that changing the failed test from `y % 2 == 0` to `y % 2 == 1` results in test failures on 64-bit OS as well.
It appears that when using `#[map]` and `#[filter]` together, code like `Just(..).prop_filter(..)` is generated, and an error occurs when the `Just` value doesn't match the filter.

To fix this issue, I'd like to change the target of `prop_filter` application from `Just` to the following strategies:

- The strategy that generates the value passed to `Just` (when `#[map]` references a single field)
- The final strategy that combines all field strategies (when `#[map]` references multiple fields)

**Comment by nc7s** [comment]
Thanks! I'll be cherry-picking it since it's not released yet.

**Comment by nc7s** [comment]
... OK, I'll admit it's harder than I thought. Would you mind cutting a new release? So I can leverage our packaging tooling and save some handwork on patches ;)

**Comment by frozenlib** [comment]
I have now published version 0.4.0, which includes this fix.

**Comment by nc7s** [comment]
Great, thanks!
